Add "Open in External Editor" function

jabcross opened the pull request:
Add the option to open a specific file in the external editor.
It's found in the script editor's File menu.

Had to extend ScriptTextEditor to provide line and
column as a Vector2i to preserve cursor position.

(The shortcut in the picture is not there by default, I assigned it locally to test it.)

![Screenshot_2019-10-08_17-53-45](https://user-images.githubusercontent.com/6402237/66432570-c4873c80-e9f4-11e9-9702-e7b244efbe97.png)

jabcross commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -2016,6 +2023,66 @@ Error ScriptEditor::_save_text_file(Ref<TextFile> p_text_file, const String &p_p
+	WARN_PRINT("Couldn't open external text editor, using internal");
+	return true;
+}
+
```
Just extracted this block into a function, no big changes here.

jabcross commented:
Fixed formatting issues.

Paulb23 commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -2045,62 +2112,7 @@ bool ScriptEditor::edit(const RES &p_resource, int p_line, int p_col, bool p_gra
-		if (err == OK)
-			return false;
-		WARN_PRINT("Couldn't open external text editor, using internal");
+		open_script_in_external_editor(p_resource, p_line, p_col);
```
The return needs to be handed to keep the current behavior.

jabcross replied:
Yes, but only if false, right? I added an if-statement.

Paulb23 commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1339,6 +1339,17 @@ void ScriptTextEditor::_edit_option(int p_op) {
 	}
 }
 
+Vector2i ScriptTextEditor::get_cursor_position() {
```
Rather then adding a new method, it would be beneficial to use `get_edit_state()` instead, this way it's agnostic to the current editor. The state has `state["row"]` and `state["column"]` for the caret position.

jabcross replied:
Interesting. Didn't know about that.

I'll take a look asap.

Paulb23 reviewed (changes requested):
Couple notes below, 

You've also squashed into the commit at the HEAD of master when you rebased, rather than your own commit.

jabcross commented:
Yeah, something weird happened when I force-pushed. Gotta look into it.

jabcross committed:
Add "Edit in External Editor" function

Add the option to open a specific file in the external editor.
It's found in the script editor's File menu.

jabcross commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -1018,6 +1018,13 @@ void ScriptEditor::_menu_option(int p_option) {
+			Vector2i position = editor->get_cursor_position();
+			// x is line, y is column
+			open_script_in_external_editor(_get_current_script(), position.x, position.y);
+		} break;
```
So it does! Thanks.

jabcross commented:
So, how do I mark that I've done the changes?

Paulb23 commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -2016,6 +2027,66 @@ Error ScriptEditor::_save_text_file(Ref<TextFile> p_text_file, const String &p_p
+	Error err = OS::get_singleton()->execute(path, args, false);
+	if (err == OK)
+		return false;
+	WARN_PRINT("Couldn't open external text editor, using internal");
```
It should return true on success rather then false.

I'm unsure if this print would be better off in the caller, allowing it to be changed for the file menu, as the script will already be open internaly?

jabcross replied:
That's what the old behavior did, tho. Returned false if err == ok.

I agree that the print should be moved, but shouldn't that be in a different PR?

Paulb23 replied:
So long as the edit function still returns correctly it doesn't matter. Calling `open_script_in_external_editor` and it returning false when its successfully opened doesn't feel quite right.

As for the print, there no problem in moving it within this PR.

Paulb23 commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -1196,6 +1196,17 @@ void ScriptEditor::_menu_option(int p_option) {
 			} break;
+			case FILE_OPEN_EXTERNAL: {
+				ScriptTextEditor *editor = Object::cast_to<ScriptTextEditor>(_get_current_editor());
+				if (!editor) return;
```
No need to cast or null check here as its already casted and checked above to "current".

Paulb23 replied:
Just this small change, then it's all good.

Paulb23 commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -2044,63 +2115,9 @@ bool ScriptEditor::edit(const RES &p_resource, int p_line, int p_col, bool p_gra
 			p_resource->get_path().is_resource_file() &&
 			p_resource->get_class_name() != StringName("VisualScript") &&
```
These two checks:
`p_resource->get_path().is_resource_file() && p_resource->get_class_name() != StringName("VisualScript")` 
ought be be moved into the new method `open_script_in_external_editor`.

jabcross replied:
Agreed.

Paulb23 reviewed (changes requested):
A few minor nitpicks below otherwise for the most part looks good.
<br />
> So, how do I mark that I've done the changes?

Just leaving a comment is fine, anybody watching will get notified.

jabcross commented:
I remembered this and made the requested changes.

jabcross commented:
Changed name of function to "Edit in external editor" instead of "Open in external editor" to differentiate between commands that open a different file. Also moved it to below "Show in Filesystem", to be with commands that operate on the current file.

jabcross commented:
Moved warn to the caller (forgot to do it before)

Paulb23 commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -2058,65 +2138,10 @@ bool ScriptEditor::edit(const RES &p_resource, int p_line, int p_col, bool p_gra
-		if (!has_file_flag) {
-			args.push_back(script_path);
+		if (open_script_in_external_editor(p_resource, p_line, p_col)) {
+			return true;
```
Should return `false` here.

jabcross replied:
Why? Opening the file was a success. `edit()` should return true, no? If opening the file fails, then we proceed to try to open in in the internal editor, who will then decide what to return.

Paulb23 replied:
`edit()` should only return true if it was successful in opening the file internally. eg.

```
if (open_script_in_external_editor(p_resource, p_line, p_col)) {
    return false;
}

/* Internal opening code */
```

jabcross replied:
Doesn't seem very clear to me that that's the expected behavior, so I added some documentation clearing it up.

jabcross replied:
By the way, the only place I can find that uses this return value is in `scipt_goto_method`, and the only place that uses _that_ return value is in `connections_dialog.cpp:775`. As I understand, it makes sure that the script is open in the internal editor before it selects a method connected via GUI.

Would this feature be desirable in the external editor as well?

Paulb23 reviewed (changes requested):
One item still to change from the last review.

Otherwise just a return change and it should be good to go.

jabcross commented:
Changes done.

aaronfranke commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -2033,7 +2044,82 @@ Error ScriptEditor::_save_text_file(Ref<TextFile> p_text_file, const String &p_p
 
+bool ScriptEditor::open_script_in_external_editor(const RES &p_resource, int p_line, int p_col) {
+
+	if (!p_resource->get_path().is_resource_file()) {
```
Empty lines after the opening brace are no longer allowed in the latest master. `clang-format` should be able to fix this for you.

Anutrix commented:
@jabcross Request you to rebase and format the files. Seems like a really useful feature.

KoBeWi commented:
Opening files in external program was added to FileSystem dock in #68489.
So what this PR does can now be achieved, in 2 steps: File -> Show in Filesystem --> Right Click -> Open in External Program.

AThousandShips commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -2033,7 +2044,82 @@ Error ScriptEditor::_save_text_file(Ref<TextFile> p_text_file, const String &p_p
+					has_file_flag = true;
+				}
+
+				// do path replacement here, else there will be issues with spaces and quotes
```
```suggestion
				// Do path replacement here, else there will be issues with spaces and quotes.
```
Code style, comments should start with a capital and end with a period (Applies to all code modified, even if it is just copied, to make sure we follow it)

fire commented:
@KoBeWi Does that mean this pr is superseded by new editor features or is there is still some value to adding this.

KoBeWi commented:
Well I think this PR is rather redundant. There is no reason to open a script in the script editor and then open it in the external editor. Since we have an option to open file in external editor now, the feature added here needs to be justified with a proposal.

fire commented:
Since we can open the files in the external editor now, we need to discuss the feature added here in a proposal. The PR also seems redundant to me.

Do you think we should close? If the mood changes, we can revisit the feature enhancement.

KoBeWi commented:
There hasn't been activity from the author, there is no proposal and the usefulness is not obvious. I agree it can be closed at this point.

Thanks for the contribution nonetheless.
